### PR TITLE
add option to enable query scheduler in helm chart

### PR DESCRIPTION
**Pull request by sysedwinistrator**
#### What this PR does
This adds an option to the mimir-distributed helm chart to enable the query-scheduler component. The query-scheduler is required to scale out the query-frontend.

#### Which issue(s) this PR fixes or relates to

The lack of the feature has been reported in grafana/mimir#2593.

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:32]: You should be able to rebase onto main and use 
```
        {{- include "mimir.podAnnotations" (dict "ctx" . "component" "query-scheduler") | nindent 8 }}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-servmon.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-servmon.yaml:9]: Can you make this default to the release namespace?

```suggestion
  namespace: {{ . | default $.Release.Namespace }}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:118]: You may need to rebase onto main for this to work. But basically it can be a secret or a configmap now and there's a handy named template that knows which one it should be.

```suggestion
        - name: config
          {{- include "mimir.configVolume" . | nindent 10 }}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:72]: Do we need this and the `volume` below? AFAIK the scheduler [doesn't store anything on disk](https://grafana.com/docs/mimir/latest/operators-guide/architecture/components/query-scheduler/).

**Review by dimitarvdimitrov — COMMENTED** [review]
looks good! Thank you for contributing this. Have you managed to test this already?

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (sysedwinistrator) [reply operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml]: done

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (sysedwinistrator) [reply operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml]: Removed it.

**Review by dimitarvdimitrov — COMMENTED** [review]
LGTM in terms of helm templates!

Can you also add a CHANGELOG entry to `operations/helm/charts/mimir-distributed/CHANGELOG.md`?

Can you also sing the CLA? You can click [here](https://cla-assistant.io/grafana/mimir?pullRequest=2087)

Can you also enable this in the OSS test `operations/helm/charts/mimir-distributed/ci/test-oss-values.yaml`? This will give us a rendered helm chart. We use this to detect changes to the resulting k8s resources when the helm templates change. After adding it you need to run `make build-helm-tests` in the root of the repo.

**Comment by sysedwinistrator** [comment]
> I commented the above, but forgot to ask you if you have plans to add anything else to the PR? It is hard to tell since the description is empty and it is still a draft PR.

The reporter of grafana/helm-charts#1404 also requests to add an option to enable query sharding *and* enable it by default. But it looks to me like the intent of the recently implemented structuredConfig option in #2100 is to make it easier to the patch the mimir config instead of having to add global switches for every feature.  
Since enabling query sharding can be done by setting a single option (`query-frontend.parallelize-shardable-queries=true`), I don't think it makes sense to add an option to the values.yaml which has to be templated somewhere else in the config. 

On the other hand, I think it might make sense to set `query-frontend.parallelize-shardable-queries` to `true` in the provided mimir config even though the default in the mimir documentation is `false`, because it is reasonable to assume a large deployment for users of this helm chart. And of course it can be easily be disabled via the `structuredConfig` parameter.

**Comment by dimitarvdimitrov** [comment]
The jsonnet also doesn't enable them by default. I'm on favour of having both of them "on" by default because I don't think they have considerable downsides. I commented on the linked issue in helm-charts

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:861]: I would make this 2 replicas. The [docs](https://grafana.com/docs/mimir/latest/operators-guide/architecture/components/query-scheduler/#operational-considerations) recommend running at least two replicas.   Having only one replica combined with the `maxUnavailable:1, maxSurge:0` means that on upgrades there will be downtime. 

```suggestion
  replicas: 2
```

**Comment by sysedwinistrator** [comment]
> now that we merged #2431 I think we should also make the query scheduler enabled by default in the chart. Can you make that change, please?

Done :)

File: CHANGELOG.md
Comment (dimitarvdimitrov) [inline CHANGELOG.md:26]: Can you move this CHANGELOG entry to `operations/helm/charts/mimir-distributed/CHANGELOG.md`?

May also be worth rebasing onto `grafana:main`

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:895]:
```suggestion
                    - query-scheduler
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:116]: otherwise my query-scheduler starts crashlooping

```suggestion
{{- end }}
        - name: storage
          emptyDir: {}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:63]:
```suggestion
              mountPath: /etc/mimir
            - name: storage
              mountPath: /data
              {{- if .Values.query_scheduler.persistence.subPath }}
              subPath: {{ .Values.query_scheduler.persistence.subPath }}
              {{- end }}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:25]: all resources should have a namespace
```suggestion
        {{- include "mimir.podAnnotations" (dict "ctx" . "component" "query-scheduler") | nindent 8 }}
      namespace: {{ .Release.Namespace | quote }}
```

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:908]:
```suggestion
  securityContext: {}
  
  # -- The SecurityContext for query_scheduler containers
  containerSecurityContext:
    readOnlyRootFilesystem: true
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:78]: Following what other components have
```suggestion
            {{- toYaml .Values.query_scheduler.containerSecurityContext | nindent 12 }}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-servmon.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-servmon.yaml:1]: You can now replace the contents of this file with 
```
{{- include "mimir.lib.serviceMonitor" (dict "ctx" $ "component" "query-scheduler") }}
```

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-pdb.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-pdb.yaml:7]: can you add namespace here too?

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-svc.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-svc.yaml:4]: can you add namespace to the metadata here too, please?

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-svc-headless.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-svc-headless.yaml:4]: Can you add namespace here too please?

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:120]: There is already a `frontend_worker` block further down. Can you move the changes here there?

**Review by dimitarvdimitrov — COMMENTED** [review]
left some more comments. thank you for your patience with this PR. I am planning on spending more time in the near days to get it merged soon. Feel free to message me [in the community slack](https://join.slack.com/t/grafana/shared_invite/zt-1dd7vh6o0-v3rEAKngEnRU9iAduBaSKw) if you have any quesitons

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-pdb.yaml
Comment (krajorama) [reply operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-pdb.yaml:7]: we've merged a PR where this whole file can be simply generated with a named template, see https://github.com/grafana/mimir/pull/2393

**Comment by dimitarvdimitrov** [comment]
hey @sysedwinistrator do you have time to work on this? If not, I'll be happy to take it over.

**Comment by dimitarvdimitrov** [comment]
can you also run `make build-helm-tests` locally from the root of the repo and commit and push the changes?

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:31]:
```suggestion
      {{- if .Values.query_scheduler.priorityClassName }}
      priorityClassName: {{ .Values.query_scheduler.priorityClassName }}
      {{- end }}
```

Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:31]: also this field does not appear in the default values.yaml. Can you add it and add a short docstring above it?

**Review by dimitarvdimitrov — COMMENTED** [review]
I think we're almost there! :) left a few more comments

**Comment by luistilingue** [comment]
I'm very excited to use that new chart with scheduler support \o/

**Comment by dimitarvdimitrov** [comment]
thanks for addressing my feedback @sysedwinistrator. I opened a PR against your branch to update the rendered manifests https://github.com/sysedwinistrator/mimir/pull/1 Can you merge that first to get the CI passing and then I can merge _this_ PR?

**Comment by sysedwinistrator** [comment]
Sorry, I just realized none of the comments I left previously were actually visible because of GitHub's feature (let's call it that..) of aggregating comments in a Review.

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml:26]: yes and I saw you did that 👍

File: operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/query-scheduler/query-scheduler-dep.yaml]: appreciated! :)

**Comment by dimitarvdimitrov** [comment]
The helm linting is failing. The PR needs one more rebase and `make build-helm-tests`. I think it will be faster if you rebase @sysedwinistrator and I can submit another PR to your branch for rebuilding the tests.

**Comment by sysedwinistrator** [comment]
FYI I also pushed c9a9dcb which was on my local branch but I forgot to push previously. It fixes the indent of the if condition around priorityClassName like you suggested for the other deployment templates as well.

**Comment by dimitarvdimitrov** [comment]
thanks, I opened https://github.com/sysedwinistrator/mimir/pull/2 with the rebuilt tests

**Comment by dimitarvdimitrov** [comment]
The CI is failing. I can take a look at it tomorrow

**Comment by dimitarvdimitrov** [comment]
Hey, turns out that adding two query-schedulers to CI runs pushed the resources of the runner to the limits. I opened https://github.com/grafana/mimir/pull/2735 to remove resource requirements in helm CI. That should unblock your PR. You will also need to rebase once that is merged. I can tag you here once 2735 is merged

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, thank you!

**Pull request merged**
